samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,718 @@
+	uint64_t value 		= 0;
+	int res			= 0;
+
+	if (!dev || !fpga_perf) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}	
+	if (fpga_perf_check_and_lock(fpga_perf)) {
+		OPAE_ERR("Failed to lock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+
+	struct udev_list_entry *attrs = udev_device_get_sysattr_list_entry(dev);
+	struct udev_list_entry *le = NULL;
+	fpga_perf->num_format = 0;
+	
+	udev_list_entry_foreach(le, attrs) {
+		const char *attr = udev_list_entry_get_name(le);
+		if (strstr(attr, "format"))
+			fpga_perf->num_format++;
+	}
+	fpga_perf->format_type = calloc(fpga_perf->num_format,
+			sizeof(perf_format_type));			
+	if (!fpga_perf->format_type) {
+		fpga_perf->num_format = 0;
+		OPAE_ERR("Failed to allocate Memory");
+		ret = FPGA_NO_MEMORY;
+		goto out;
+	}
+	udev_list_entry_foreach(le, attrs) {
+		const char *attr = udev_list_entry_get_name(le);
+		if (strstr(attr, "format")) {
+			reg_res = regcomp(&re, PERF_CONFIG_PATTERN,
+					REG_EXTENDED | REG_ICASE);
+			if (reg_res) {
+				OPAE_ERR("Error compiling regex");
+				ret = FPGA_EXCEPTION;
+				goto out;
+			}
+			reg_res = regexec(&re, udev_device_get_sysattr_value(dev, attr),
+					4, matches, 0);
+			if (reg_res) {
+				regerror(reg_res, &re, err, sizeof(err));
+				OPAE_ERR("Error executing regex: %s", err);
+				ret = FPGA_EXCEPTION;
+				goto out;
+			} else {
+				const char *attr_value = udev_device_get_sysattr_value(dev, attr);
+				PARSE_MATCH_INT(attr_value, matches[1].rm_so, value, 10);
+				fpga_perf->format_type[loop].shift = value;
+				if (snprintf(fpga_perf->format_type[loop].format_name,
+					sizeof(fpga_perf->format_type[loop].format_name),
+					"%s", (strstr(attr, "/")+1)) < 0) {
+					OPAE_ERR("snprintf buffer overflow");
+					ret = FPGA_EXCEPTION;
+					goto out;
+				}
+				loop++;
+			}
+		}
+	}
+	if (opae_mutex_unlock(res, &fpga_perf->lock)) {
+		OPAE_ERR("Failed to unlock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+out:
+	opae_mutex_unlock(res, &fpga_perf->lock);
+	return ret;
+}
+
+/* parse the events for the particular device directory */
+fpga_result parse_perf_event(struct udev_device *dev, fpga_perf_counter *fpga_perf)
+{	
+	fpga_result ret		= FPGA_OK;
+	regex_t re;
+	regmatch_t matches[4] 	= { {0} };
+	char err[128] 		= { 0 };
+	int reg_res		= 0;
+	uint64_t loop 		= 0;
+	uint64_t inner_loop 	= 0;
+	int res			= 0;
+
+	if (!dev || !fpga_perf) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+	if (fpga_perf_check_and_lock(fpga_perf)) {
+		OPAE_ERR("Failed to lock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	struct udev_list_entry *attrs = udev_device_get_sysattr_list_entry(dev);
+	struct udev_list_entry *le = NULL;
+	fpga_perf->num_perf_events = 0;
+	
+	udev_list_entry_foreach(le, attrs) {
+		const char *attr = udev_list_entry_get_name(le);
+		if (strstr(attr, "events"))
+			fpga_perf->num_perf_events++;
+	}
+	fpga_perf->perf_events = calloc(fpga_perf->num_perf_events,
```

Comment:
I this this could be a memory leak if we call parse_perf_event multiple times using the same fpga_perf_counter structure.